### PR TITLE
keepassxc-cli: Implement attachment handling

### PR DESCRIPTION
**Pull request by andreblanke**
This pull request implements three new commands and strives to address #4462:

- `attachment-export database entry name` which exports the contents of an attachment inside a database entry to the command line
- `attachment-import [-f] database entry name attachment` which imports a file `attachment` into a database entry under a given name. An existing attachment with the same name may be overridden if the `-f` option is specified
- `attachment-rm database entry name` which removes the attachment with the provided name from the entry inside the database

## Design decisions

The linked issue suggests modifications to the `add` and `edit` commands instead of introducing new commands like I have done but I felt having dedicated commands for it was a cleaner approach then adding additional functionality to `add` and `edit`, even though I am a bit unsure about it.
A `getattachment` or `attachment-export` command would be needed anyway and this way there exists symmetry between `attachment-export` and `attachment-import`.

For consistency with `export`, `attachment-export` outputs the contents of an attachment to the command line instead of taking an additional file argument.

I am unsure whether a command like `attachment-mv` to rename an attachment or to move it from one entry to another would be a worthwhile addition, as that use case can be covered using a combination of the commands implemented in this pull request.

Additionally, the `name` argument of `attachment-import` could be made optional, in which case it would default to the name of the attachment file. In that case switching the order of arguments to `attachment-import [-f] database entry attachment [name]` might also make sense.

## Testing strategy

A unit test for each of the new commands has been added in `TestCli.cpp`.
Along with that I introduced `tests/data/Attachment.txt` used as a test attachment and added the `Sample attachment.txt` (with content "Sample content\n") to the `/Sample Entry` in `NewDatabase.kdbx`.

## Type of change

- ✅ New feature (change that adds functionality)

### CODE REVIEW

File: src/cli/AttachmentExport.cpp
Comment (Colfenor) [inline src/cli/AttachmentExport.cpp:41]: Hi ! (I'm not sure if I'm allowed to review), however I want to point out will have to null/sanity check your `database` & `parser` objects right after passing them to your functions or you will get segmentation-faults in case of misinitialization.

Comment (andreblanke) [reply src/cli/AttachmentExport.cpp:41]: Hey there! That would surely be the most defensive approach to it but none of the other commands I have checked do those checks which is why I figured it would not be needed. For now I'll wait and see what others have to say about it.

Comment (Colfenor) [reply src/cli/AttachmentExport.cpp:41]: Just because in other commands the asserts/checks were omitted it doesn't necessarily mean that this is the right way to do it. (I'd suggest fixing them aswell)

What would you suppose should happen in a case of a null object ? You then have undefined behaviour and not handeld an error case properly. Another option could be to pass the function arguments as references.

Comment (andreblanke) [reply src/cli/AttachmentExport.cpp:41]: But a null database or parser object would never reach an implementation of `DatabaseCommand::executeWithDatabase` because `DatabaseCommand::execute`, which later invokes that method, already does the null checking for us [here](https://github.com/keepassxreboot/keepassxc/blob/fd3cc7e8c3d012d74d649773eb2236bd06f12fbc/src/cli/DatabaseCommand.cpp#L40-L42) and [here](https://github.com/keepassxreboot/keepassxc/blob/fd3cc7e8c3d012d74d649773eb2236bd06f12fbc/src/cli/DatabaseCommand.cpp#L61-L63), so we would only be duplicating code.

Comment (Colfenor) [reply src/cli/AttachmentExport.cpp:41]: > 
> 
> But a null database or parser object would never reach an implementation of `DatabaseCommand::executeWithDatabase` because `DatabaseCommand::execute`, which later invokes that method, already does the null checking for us [here](https://github.com/keepassxreboot/keepassxc/blob/fd3cc7e8c3d012d74d649773eb2236bd06f12fbc/src/cli/DatabaseCommand.cpp#L40-L42) and [here](https://github.com/keepassxreboot/keepassxc/blob/fd3cc7e8c3d012d74d649773eb2236bd06f12fbc/src/cli/DatabaseCommand.cpp#L61-L63), so we would only be duplicating code.

You are right in this specific case the caller has taken care of the null check, but IMO null checking should be in the callee (the called function) were the variable goes into scope. 
This defeats the case that if someone else calls your function & _does not null check before_ for whatever reason (intentionally or unintentionally) you'd drive into the same problem again. 

So for every function call of yours the caller would have to write a designated null check.

Comment (louib) [reply src/cli/AttachmentExport.cpp:41]: > I'm not sure if I'm allowed to review

@Colfenor Of course you are!

Comment (andreblanke) [reply src/cli/AttachmentExport.cpp:41]: @Colfenor While I agree with you I think the changes you talk about should be part of a different pull request that fixes all existing DatabaseCommands. An alternative would be marking `executeWithDatabase` as protected to prevent invocation from outside classes.

Comment (Colfenor) [reply src/cli/AttachmentExport.cpp:41]: We'll if you agree, may I ask what is stopping you from fixing them in your PR ? (or at least sanitize the arguments (null-check) with your three new commands)

Comment (andreblanke) [reply src/cli/AttachmentExport.cpp:41]: Sorry for the delayed response. To answer your question: Because I really don't think of it as a change which is too important. It certainly doesn't have downsides to it but there is null safety in place already (albeit not directly inside the method) and I think it would make more sense for a single commit to fix all occurrences of it, which you could tackle once this pull request has been merged.
I'd prefer dropping the null debate for now if you don't mind but I would of course still be happy about additional feedback regarding the code I wrote.

**Comment by zarelit** [comment]
This PR is almost ready and it brings something towards the CLI as a full alternative to the GUI. I use attachments extensively and in the CLI it appears like they don't even exist. 

Now the PR is stuck on something that can be debated and should be debated but it definitely looks broader and deeper than the scope of the PR itself.

By reading the comments it looks like in this particular instance both the solutions might be safe in the present and only become problematic as the development goes on and the caller changes or new code does not respect the same style.

Is there someone willing to unstuck this? Maybe @louib?

File: src/cli/AttachmentExport.cpp
Comment (louib) [reply src/cli/AttachmentExport.cpp]: @andreblanke I agree that we can defer that specific (null checking) change to a later, dedicated PR. It should indeed already be done before calling `executeWithDatabase`.

File: docs/man/keepassxc-cli.1.adoc
Comment (louib) [inline docs/man/keepassxc-cli.1.adoc]: @andreblanke maybe `remove an attachment from an entry`?

File: docs/man/keepassxc-cli.1.adoc
Comment (louib) [inline docs/man/keepassxc-cli.1.adoc]: @andreblanke maybe we should rename `name` to `attachment_name`, it feels more explicit. Same comment goes for all 3 commands.

File: docs/man/keepassxc-cli.1.adoc
Comment (louib) [inline docs/man/keepassxc-cli.1.adoc]: @andreblanke I think we can remove `of an entry` in that case, since it's obvious from the usage documentation.

File: src/cli/AttachmentExport.h
Comment (louib) [inline src/cli/AttachmentExport.h]: @andreblanke namespace should be `ATTACHMENTEXPORT`

File: tests/TestCli.cpp
Comment (louib) [inline tests/TestCli.cpp]: @andreblanke `attachmentPath` maybe? otherwise it's not clear if it's the path of the content of the attachment.

**Review by louib — CHANGES_REQUESTED** [review]
@andreblanke Added a few comments, but overall I think this PR is really close to be ready to merge. The tests are looking good. Thx for the PR! :pray:

File: docs/man/keepassxc-cli.1.adoc
Comment (droidmonkey) [reply docs/man/keepassxc-cli.1.adoc]: Removes the named attachment from an entry?

**Comment by droidmonkey** [comment]
I personally don't like that we are introducing new named commands. I would much rather handle this through entry editing with switches for attachment handling. 

`keepassxc-cli edit --attachment-add _attachmentfile_ _database_ _entry_`

`keepassxc-cli show --attachment _attachmentname_ _database_ _entry_`

`keepassxc-cli edit --attachment-rm _attachmentname_ _database_ _entry_`

This is the defined workflow with our cli, by adding new commands you create new workflows and that is generally frowned upon.

**Comment by louib** [comment]
@droidmonkey the main problem that I see with jamming all the attachment commands in `add` and `edit` is that, when you add an attachment, you require 2 params: the name and the path. Would you use 2 new options just for that? Or encode both in the `--attachment-add` param?  I feel like the attachments are specific enough to justify a new set of commands, and I'm not a fan of hiding sub-commands into other commands using parameters. I'd rather have a lot of small, focused commands than a few overloaded commands. Also, if I remember clearly, the `add` and `edit` commands are still missing some other fields, so they would become even more bloated.

Whichever decision we make, I think we should also update the `Show` command in this PR to display the attachment names of the entry, otherwise I believe there's no way to see the attachments of an entry from the CLI.

**Comment by droidmonkey** [comment]
Can we just add an "attachment" command and use add/edit/get subcommands?

**Comment by louib** [comment]
@droidmonkey that could be an option, although it would be inconsistent with what we have for the entry commands, the group commands and the db commands, which have dedicated commands for every action.

**Comment by droidmonkey** [comment]
Yah I agree, blah... this just sets up bad precedent moving forward for other features. We need to be careful of command creep... already have that problem!

**Comment by andreblanke** [comment]
I'm sorry for this pull request becoming stale, partially due to the discussion. In hindsight I should have just added the additional null safety. However, I mainly remember not being able to resolve the build failures and pausing development due to that.

@louib Thanks for the feedback! I will try to implement the changes you suggested in the next few days. I unfortunately don't have too much time on my hands at the moment but will try to get to it soon.

Concerning the command design: I tried to stay consistent but agree that an `attachment` command with the various subcommands would definitely be the better choice for the future, though inconsistent.

**Comment by louib** [comment]
@andreblanke let me know when you need another review!

**Review by louib — CHANGES_REQUESTED** [review]
Almost there!

A few things:
* You can format the codebase using `make format`. This should fix the formatting CI job.
* We should add the attachments in the output of the `show` command, otherwise there's no way to see the attachments of an entry from the CLI.
* There's a conflict in the `NewDatabase` file, you'll probably need to rebase your branch against `develop` and fix the conflict. It'll probably be easier to start from the version on `develop` and re-apply your changes on it.

**Comment by droidmonkey** [comment]
Could also merge your version of new database into the develop version using keepassxc merge feature 😃

File: src/cli/Show.cpp
Comment (louib) [inline src/cli/Show.cpp]: I'm not sure we should print that if there are no attachments. But users might also think that the feature is not implemented at all. Maybe we could print a specific string, like `No attachments`?

Comment (andreblanke) [reply src/cli/Show.cpp]: Thanks for pointing that out or else I'd have probably forgotten it. I was going to get back to you for more specific details regarding the formatting anyway. Printing `No attachments` sounds like a good idea.

**Comment by louib** [comment]
@andreblanke the only thing remaining would be to add some tests to the `show` command test case for entries with and without attachments. You'll need to configure the project with `WITH_GUI_TESTS` to run the CLI unit tests locally. It's a bit counterintuitive, I will be fixing that soon.

**Comment by andreblanke** [comment]
I introduced the `--show-attachments` option to the `show` to avoid printing attachment information when only specific attributes are requested via the `--attributes` flag. If you'd like me to handle it any differently please let me know.

Other than that the tests should be good to go with the exception of `TestGui::testAutoreloadDatabase` and `TestGui::testEditEntry` which are most likely failing due to the changes done to `NewDatabase.kdbx`, though I'm not to sure how to fix those at the moment but I haven't looked into it too far either.

File: src/cli/AttachmentRemove.cpp
Comment (droidmonkey) [inline src/cli/AttachmentRemove.cpp]: This import is not necessary

File: src/cli/AttachmentExport.h
Comment (droidmonkey) [inline src/cli/AttachmentExport.h]: This import should be in the cpp only

File: src/cli/AttachmentExport.cpp
Comment (droidmonkey) [inline src/cli/AttachmentExport.cpp]: These imports are not necessary, already in the header

File: src/cli/AttachmentExport.cpp
Comment (droidmonkey) [inline src/cli/AttachmentExport.cpp]: I would prefer some way to also optionally specify the output file name instead of always dumping to stdout.

**Review by droidmonkey — CHANGES_REQUESTED** [review]

File: src/cli/AttachmentExport.cpp
Comment (andreblanke) [reply src/cli/AttachmentExport.cpp]: I decided to entirely remove dumping to stdout when using `attachment-export` for the following reasons:
1. Symmetry with `attachment-import` which is only able to import from a file, not from stdin.
2. Binary output might mess up the terminal.
3. I found no good way to output the attachment which could both be checked by the tests using `m_stdout->readAll()` but also actually displayed the content to the user. All solutions I tried so far only allowed one of those things (at least without including some kind of workaround like `m_fout` that was used previously in `AttachmentExport`, but is now removed).

Comment (droidmonkey) [reply src/cli/AttachmentExport.cpp]: I concur with this and it's consistent with how the gui handles attachments... as files.

Comment (droidmonkey) [reply src/cli/AttachmentExport.cpp]: Thinking this through... there might be value in a `--stdout` flag on attachment-export for people who want to pipe the contents of an attachment into another command. When that flag is issued then the export_file positional parameter is ignored/optional. This is coming from #6947

**Comment by louib** [comment]
@andreblanke everything is looking good, except for the tests failing. Might need a rebase on `develop` as well.

**Comment by andreblanke** [comment]
Alright. I'm a bit busy at the moment but I'll try to get around to it on the weekend. While I'm at it I'll look into adding `--stdout` as well.

**Comment by droidmonkey** [comment]
I squashed all the commits, fixed the tests, and found a typo too! This is ready for merge after CI completes.

**Review by droidmonkey — APPROVED** [review]

**Comment by droidmonkey** [comment]
Still needs a little more work after fixing the size display in show command.

**Comment by louib** [comment]
@droidmonkey the CI jobs do not look happy :(

**Comment by louib** [comment]
@droidmonkey some tests are failing, and apparently it's related to entry history?
```
FAIL!  : TestGui::testAutoreloadDatabase() Compared values are not the same
   Actual   (entry->historyItems().size()): 10
   Expected (++editCount)                 : 11
   Loc: [/opt/buildagent/work/c401303cba1b4098/tests/gui/TestGui.cpp(490)]
QDEBUG : TestGui::testAutoreloadDatabase() Merge Sample Entry/Sample Entry_test with alien on top under NewDatabase
QWARN  : TestGui::testAutoreloadDatabase() Inconsistent history entry of Sample Entry[9f4544c2ab00c74a8a1a6eaf26cf57e9] at 2021-11-04 11-08-32-000 contains conflicting changes - conflict resolution may lose data!
PASS   : TestGui::testTabs()
FAIL!  : TestGui::testEditEntry() Compared values are not the same
   Actual   (entry->historyItems().size()): 10
   Expected (++editCount)                 : 11
   Loc: [/opt/buildagent/work/c401303cba1b4098/tests/gui/TestGui.cpp(490)]
```

**Comment by droidmonkey** [comment]
Yah I have no idea what is going on, could be because the entry was edited in NewDatabase.kdbx

**Comment by andreblanke** [comment]
The GUI tests were exactly the ones I didn't know how to fix. Glad I wasn't the only one confused by them though. Anyway, thanks @droidmonkey for finishing off the PR. It was a pleasure working with you two and I appreciate the feedback that you gave me regarding my commits

**Pull request merged**
